server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/fileformats/meta/StarTreeMetadata.java

```diff
@@ -0,0 +1,358 @@
+import org.opensearch.index.mapper.CompositeMappedFieldType;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Holds the associated metadata for the building of star-tree.
+ *
+ * @opensearch.experimental
+ */
+public class StarTreeMetadata extends CompositeIndexMetadata {
+    private static final Logger logger = LogManager.getLogger(StarTreeMetadata.class);
+
+    /**
+     * The index input for reading metadata from the segment file.
+     */
+    private final IndexInput meta;
+
+    /**
+     * The version of the star tree stored in the segments.
+     */
+    private final int version;
+
+    /**
+     * The number of the nodes in the respective star tree
+     */
+    private final int numberOfNodes;
+
+    /**
+     * The name of the star-tree field, used to identify the star-tree.
+     */
+    private final String starTreeFieldName;
+
+    /**
+     * The type of the star-tree field, indicating the specific implementation or version.
+     * Here, STAR_TREE field.
+     */
+    private final String starTreeFieldType;
+
+    /**
+     * List of dimension fields used in the star-tree.
+     */
+    private final List<String> dimensionFields;
+
+    /**
+     * List of metric entries, containing field names and associated metric statistic.
+     */
+    private final List<MetricEntry> metricEntries;
+
+    /**
+     * The total number of documents aggregated in this star-tree segment.
+     */
+    private final Integer segmentAggregatedDocCount;
+
+    /**
+     * The maximum number of documents allowed in a leaf node.
+     */
+    private final Integer maxLeafDocs;
+
+    /**
+     * Set of dimensions for which star node creation should be skipped.
+     */
+    private final Set<String> skipStarNodeCreationInDims;
+
+    /**
+     * The build mode used for constructing the star-tree.
+     */
+    private final StarTreeFieldConfiguration.StarTreeBuildMode starTreeBuildMode;
+
+    /**
+     * The file pointer to the start of the associated star-tree data in the (.cid) file
+     */
+    private final long dataStartFilePointer;
+
+    /**
+     * The length of the star-tree data in bytes, used for reading the correct amount of data from (.cid) file
+     */
+    private final long dataLength;
+
+    /**
+     * A star tree metadata constructor to initialize star tree metadata from the segment file (.cim) using index input.
+     *
+     * @param metaIn             an index input to read star-tree meta
+     * @param compositeFieldName name of the composite field. Here, name of the star-tree field.
+     * @param compositeFieldType type of the composite field. Here, STAR_TREE field.
+     * @param version The version of the star tree stored in the segments.
+     * @throws IOException if unable to read star-tree metadata from the file
+     */
+    public StarTreeMetadata(
```

Comment:
Whys is metadata writer not writing just the StarTreeMetadata? Why is the writer relying on specific attributes?

Reply:
We need to construct StarTreeMetadata using the attributes that we are sending from BaseStarTreeBuilder.
We read that StarTreeMetadata then during query / merges.

Reply:
This is because how it has been implemented in this PR.
for the writer as well, if there are 2 constructors which can init either using an IndexInput, or a builder pattern with actual attributes, and then write the metadata class directly.

Given you have separated write to a separate class but read still depends on the logic of the constructor of the 
 StarTreeMetadata class, we're not achieving any version agnostic behaviour either.

Reply:
I am not fully sure on this. There are two ctors in the StarTreeMetadata currently. the method on line 138 is going to be used for testing in the next PR (as mentioned in the javadoc).

Now with this method, my understanding is, with the version as one of the arguments, I would be able to change or tweak the read of StarTreeMetadata. Also, this class is implemented to read only from the file. 

What do you think?

Reply:
Added an issue to track this in a separate PR:
https://github.com/opensearch-project/OpenSearch/issues/15467